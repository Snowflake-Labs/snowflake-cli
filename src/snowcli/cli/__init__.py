from __future__ import annotations

import logging
import pkgutil
import sys
import typer
from collections.abc import Container
from pathlib import Path

from snowcli import __about__
from snowcli.cli import (
    connection,
    render,
    sql,
    streamlit,
)
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.common.snow_cli_global_context import (
    snow_cli_global_context_manager,
    SnowCliGlobalContext,
)
from snowcli.cli.main.snow_cli_main_typer import SnowCliMainTyper
from snowcli.cli import loggers
from snowcli.cli.snowpark import app as snowpark_app
from snowcli.config import AppConfig
from snowcli.output.formats import OutputFormat
from snowcli.connection_config import ConnectionConfigs

app = SnowCliMainTyper()
log = logging.getLogger(__name__)


def _version_callback(value: bool):
    if value:
        typer.echo(f"SnowCLI Version: {__about__.VERSION}")
        raise typer.Exit()


def setup_global_context(debug: bool):
    def modifications(context: SnowCliGlobalContext) -> SnowCliGlobalContext:
        context.enable_tracebacks = debug
        return context

    snow_cli_global_context_manager.update_global_context(modifications)


@app.command()
def login(
    snowsql_config_path: Path = typer.Option(
        "~/.snowsql/config",
        "--config",
        "-c",
        prompt="Path to Snowsql config",
        help="snowsql config file",
    ),
    connection_name: str = typer.Option(
        ...,
        "--connection",
        "-C",
        prompt="Connection name (for entry in snowsql config)",
        help="connection name from snowsql config file",
    ),
):
    """
    Select a Snowflake connection to use with SnowCLI.
    """
    if not snowsql_config_path.expanduser().exists():
        log.error(f"Path to snowsql config does not exist: {snowsql_config_path}")
        raise typer.Abort()

    connection_configs = ConnectionConfigs(snowsql_config_path)
    if not connection_configs.connection_exists(connection_name):
        log.error(
            f"Connection not found in {snowsql_config_path}: {connection_name}. "
            "You can add with `snow connection add`."
        )
        raise typer.Abort()

    cfg = AppConfig()
    cfg.config["snowsql_config_path"] = str(snowsql_config_path.expanduser())
    cfg.config["snowsql_connection_name"] = connection_name
    cfg.save()
    log.info(f"Using connection name {connection_name} in {snowsql_config_path}")
    log.info(f"Wrote {cfg.path}")


@app.command()
def configure(
    environment: str = typer.Option(
        "dev",
        "-e",
        "--environment",
        help="Name of environment (e.g. dev, prod, staging)",
    ),
    database: str = typer.Option(
        ...,
        "--database",
        prompt="Snowflake database",
        help="Snowflake database",
    ),
    schema: str = typer.Option(
        ...,
        "--schema",
        prompt="Snowflake schema",
        help="Snowflake schema",
    ),
    role: str = typer.Option(
        ...,
        "--role",
        prompt="Snowflake role",
        help="Snowflake role",
    ),
    warehouse: str = typer.Option(
        ...,
        "--warehouse",
        prompt="Snowflake warehouse",
        help="Snowflake warehouse",
    ),
):
    """
    Configure an environment to use with your Snowflake connection.
    """
    log.info(f"Configuring environment {environment}...")
    cfg = AppConfig()
    if environment in cfg.config and not typer.confirm(
        "Environment {environment} already exists. Overwrite?",
    ):
        log.error("Cancelling...")
        raise typer.Abort()

    cfg.config[environment] = {}
    cfg.config[environment]["database"] = database
    cfg.config[environment]["schema"] = schema
    cfg.config[environment]["role"] = role
    cfg.config[environment]["warehouse"] = warehouse
    cfg.save()
    log.info(f"Wrote environment {environment} to {cfg.path}")


@app.callback()
def default(
    version: bool = typer.Option(
        None,
        "--version",
        help="Prints version of the snowcli",
        callback=_version_callback,
        is_eager=True,
    ),
    output_format: OutputFormat = typer.Option(
        OutputFormat.TABLE.value,
        "--format",
        help="Specifies output format",
        case_sensitive=False,
        is_eager=True,
    ),
    configuration_file: Path = typer.Option(
        None,
        "--config-file",
        help="Specifies snowcli configuration file that should be used",
        exists=True,
        dir_okay=False,
        is_eager=True,
    ),
    verbose: bool = typer.Option(
        None,
        "--verbose",
        "-v",
        help="Print logs from level info and higher",
    ),
    debug: bool = typer.Option(
        None,
        "--debug",
        help="Print logs from level debug and higher, logs contains additional information",
    ),
) -> None:
    """
    SnowCLI - A CLI for Snowflake
    """
    loggers.create_loggers(verbose, debug)
    setup_global_context(debug=debug)


MODULE_IGNORE_SET = frozenset(("main", "procedure_coverage"))


def register_cli_typers(ignore_container: Container[str] = MODULE_IGNORE_SET) -> None:
    for _, name, _ in pkgutil.walk_packages(__path__):
        if name not in ignore_container:
            cli_app = __import__(f"{__name__}.{name}", fromlist=["_trash"])
            try:
                app.add_typer(cli_app.app, name=name)
            except AttributeError:
                # Ignore modules that don't define app global
                pass


register_cli_typers()

app.command("sql")(sql.execute_sql)
app.add_typer(snowpark_app)


if __name__ == "__main__":
    app()

if getattr(sys, "frozen", False):
    app(sys.argv[1:])
