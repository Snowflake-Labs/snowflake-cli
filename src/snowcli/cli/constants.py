from __future__ import annotations

from enum import Enum

DEPLOYMENT_STAGE = "deployments"


class ObjectType(Enum):
    COMPUTE_POOL = "compute pool"
    DATABASE = "database"
    FUNCTION = "function"
    JOB = "job"
    PROCEDURE = "procedure"
    ROLE = "role"
    SCHEMA = "schema"
    SERVICE = "service"
    STREAMLIT = "streamlit"
    TABLE = "table"
    WAREHOUSE = "warehouse"
