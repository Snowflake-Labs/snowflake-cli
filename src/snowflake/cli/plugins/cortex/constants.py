from snowflake.cli.plugins.cortex.models import Model

DEFAULT_MODEL: Model = Model("snowflake-arctic")
