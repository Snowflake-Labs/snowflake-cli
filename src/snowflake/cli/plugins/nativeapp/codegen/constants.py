# All constants used across the codegen module

SNOWPARK_PROCESSOR = "snowpark"
STAGE_PREFIX = "@"
TEMP_OBJECT_NAME_PREFIX = "SNOWPARK_TEMP_"
