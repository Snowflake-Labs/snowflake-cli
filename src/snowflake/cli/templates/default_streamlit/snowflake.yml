definition_version: "2"
entities:
  my_streamlit:
    type: "streamlit"
    name: "my_dashboard"
    title: "My Streamlit Dashboard"
    stage: "my_streamlit_stage"
    query_warehouse: 'my_streamlit_warehouse'
    main_file: streamlit_app.py
    pages_dir: pages/
    artifacts:
      - streamlit_app.py
      - environment.yml
      - common/hello.py
      - pages/
