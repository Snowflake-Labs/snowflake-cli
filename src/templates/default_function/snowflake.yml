definition_version: 1
functions:
  - name: hello
    handler: "app.hello"
    signature:
      - name: "name"
        type: "string"
    returns: string
