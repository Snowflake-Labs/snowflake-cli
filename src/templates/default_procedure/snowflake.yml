definition_version: 1
procedures:
  - name: hello
    handler: "app.hello"
    signature:
      - name: "name"
        type: "string"
    returns: string
  - name: test
    handler: "app.test"
    signature: ""
    returns: string
