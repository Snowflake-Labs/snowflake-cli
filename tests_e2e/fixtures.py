import shutil
from contextlib import contextmanager
from pathlib import Path

import pytest


@pytest.fixture
def project_directory(temp_dir, test_root_path):
    @contextmanager
    def _temporary_project_directory(project_name):
        test_data_file = test_root_path / "test_data" / project_name
        shutil.copytree(test_data_file, temp_dir, dirs_exist_ok=True)
        yield Path(temp_dir)

    return _temporary_project_directory
