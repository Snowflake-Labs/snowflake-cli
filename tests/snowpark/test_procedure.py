import json
from textwrap import dedent
from unittest.mock import call

from snowcli.cli.constants import SnowparkObjectType
from snowflake.connector import ProgrammingError

from tests.testing_utils.fixtures import *


def test_deploy_function_no_procedure(runner, project_directory):
    with project_directory("empty_project"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
        )
    assert result.exit_code == 1
    assert "No procedures or functions were specified in project" in result.output


@mock.patch("snowflake.connector.connect")
@mock.patch("snowcli.cli.snowpark.commands.ObjectManager.describe")
def test_deploy_procedure(
    mock_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_describe.side_effect = ProgrammingError("does not exist or not authorized")
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ]
        )

    assert result.exit_code == 0, result.output
    mock_describe.assert_has_calls(
        [
            call(SnowparkObjectType.PROCEDURE.value, "procedureName(string)"),
            call(SnowparkObjectType.PROCEDURE.value, "test()"),
        ]
    )
    assert ctx.get_queries() == [
        "create stage if not exists deployments comment='deployments managed by snowcli'",
        f"put file://app.zip @deployments/procedurename_name_string"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace procedure procedureName(name string)
            returns string
            language python
            runtime_version=3.8
            imports=('@deployments/procedurename_name_string/app.zip')
            handler='app.hello'
            packages=()
            """
        ),
        f"put file://app.zip @deployments/test"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace procedure test()
            returns string
            language python
            runtime_version=3.8
            imports=('@deployments/test/app.zip')
            handler='app.test'
            packages=()
            """
        ),
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowcli.cli.snowpark.commands._alter_procedure_artifact")
@mock.patch("snowcli.cli.snowpark.commands.ObjectManager.describe")
def test_deploy_procedure_with_coverage(
    mock_describe,
    mock_alter_procedure_artifact,
    mock_conn,
    runner,
    mock_ctx,
    snapshot,
    temp_dir,
    project_directory,
):
    mock_alter_procedure_artifact.return_value = "snowpark_coverage.measure_coverage"

    mock_describe.side_effect = ProgrammingError("does not exist or not authorized")
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory(
        "snowpark_procedures",
        {
            "procedures": [
                {
                    "name": "foo",
                    "signature": [{"name": "name", "type": "string"}],
                    "handler": "foo.func",
                    "returns": "variant",
                }
            ]
        },
    ):
        result = runner.invoke(["snowpark", "deploy", "--install-coverage-wrapper"])

    assert result.exit_code == 0, result.output
    mock_describe.assert_has_calls(
        [call(SnowparkObjectType.PROCEDURE.value, "foo(string)")]
    )
    assert ctx.get_queries() == [
        "create stage if not exists deployments comment='deployments managed by snowcli'",
        f"put file://app.zip @deployments/foo_name_string"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace procedure foo(name string)
            returns variant
            language python
            runtime_version=3.8
            imports=('@deployments/foo_name_string/app.zip')
            handler='snowpark_coverage.measure_coverage'
            packages=('coverage')
            """
        ),
    ]


def test_coverage_wrapper_does_not_work_for_multiple_procedures(
    project_directory, runner
):
    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--install-coverage-wrapper"])
    assert result.exit_code == 1
    assert "Using coverage wrapper is currently limited" in result.output


@mock.patch("snowflake.connector.connect")
@mock.patch("snowcli.cli.snowpark.commands.ObjectManager.describe")
def test_deploy_procedure_fails_if_object_exists_and_no_replace(
    mock_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_describe.side_effect = [
        mock_cursor(
            [
                ("packages", '["foo=1.2.3", "bar>=3.0.0"]'),
                ("handler", "main.py:app"),
                ("returns", "table(variant)"),
            ],
            columns=["key", "value"],
        ),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ]
        )

    assert result.exit_code == 1
    assert "Procedure procedureName(name string) already exists." in result.output


@mock.patch("snowflake.connector.connect")
@mock.patch("snowcli.cli.snowpark.commands.ObjectManager.describe")
def test_deploy_procedure_replace_nothing_to_update(
    mock_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_describe.side_effect = [
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "app.hello"),
                ("returns", "string"),
            ],
            columns=["key", "value"],
        ),
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "app.test"),
                ("returns", "string"),
            ],
            columns=["key", "value"],
        ),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--replace", "--format", "json"])

    assert result.exit_code == 0
    assert json.loads(result.output) == [
        {
            "object": "procedureName(name string)",
            "status": "packages updated",
            "type": "procedure",
        },
        {"object": "test()", "status": "packages updated", "type": "procedure"},
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowcli.cli.snowpark.commands.ObjectManager.describe")
def test_deploy_procedure_replace_updates_single_object(
    mock_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_describe.side_effect = [
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "app.hello"),
                ("returns", "string"),
            ],
            columns=["key", "value"],
        ),
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "app.foo"),
                ("returns", "string"),
            ],
            columns=["key", "value"],
        ),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--replace", "--format", "json"])

    assert result.exit_code == 0
    assert json.loads(result.output) == [
        {
            "object": "procedureName(name string)",
            "status": "packages updated",
            "type": "procedure",
        },
        {"object": "test()", "status": "definition updated", "type": "procedure"},
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowcli.cli.snowpark.commands.ObjectManager.describe")
def test_deploy_procedure_replace_creates_missing_object(
    mock_describe,
    mock_conn,
    runner,
    mock_cursor,
    mock_ctx,
    project_directory,
):
    mock_describe.side_effect = [
        mock_cursor(
            [
                ("packages", "[]"),
                ("handler", "app.hello"),
                ("returns", "string"),
            ],
            columns=["key", "value"],
        ),
        ProgrammingError("does not exist or not authorized"),
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_procedures"):
        result = runner.invoke(["snowpark", "deploy", "--replace", "--format", "json"])

    assert result.exit_code == 0
    assert json.loads(result.output) == [
        {
            "object": "procedureName(name string)",
            "status": "packages updated",
            "type": "procedure",
        },
        {"object": "test()", "status": "created", "type": "procedure"},
    ]


@mock.patch("snowflake.connector.connect")
def test_execute_procedure(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    result = runner.invoke(
        [
            "snowpark",
            "execute",
            "procedure",
            "procedureName(42, 'string')",
        ]
    )

    assert result.exit_code == 0, result.output
    assert ctx.get_query() == "call procedureName(42, 'string')"


@mock.patch("snowcli.cli.common.project_initialisation._create_project_template")
def test_init_procedure(mock_create_project_template, runner, temp_dir):
    runner.invoke(["snowpark", "init", "my_project2"])
    mock_create_project_template.assert_called_once_with(
        "default_snowpark", project_directory="my_project2"
    )
