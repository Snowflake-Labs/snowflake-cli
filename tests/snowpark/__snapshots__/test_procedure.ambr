# serializer version: 1
# name: test_deploy_procedure_fails_if_object_exists_and_no_replace
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Following object already exists. Consider using --replace.                   │
  │  procedure: PROCEDURENAME(VARCHAR) RETURN VARCHAR                            │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
